torch_xla/csrc/index_op_util.cpp

```diff
@@ -0,0 +1,103 @@
+#include "torch_xla/csrc/index_op_util.h"
+#include <ATen/ExpandUtils.h>
+#include <ATen/Functions.h>
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "torch_xla/csrc/aten_xla_bridge.h"
+
+namespace torch_xla {
+namespace {
+
+void CheckIndexTensorTypes(at::TensorList indices) {
+  for (auto& tensor : indices) {
+    if (tensor.defined()) {
+      auto& type = tensor.type();
+      auto scalarType = type.scalarType();
+      if (scalarType != at::kLong && scalarType != at::kByte) {
+        XLA_ERROR() << "tensors used as indices must be long or byte tensors";
+      }
+    }
+  }
+}
+
+// Expands byte tensors (masks) into the equivalent indexing by LongTensors.
+// This is a version of at::native::expandByteTensors with style adjustments.
+std::vector<at::Tensor> ExpandByteTensors(const at::Tensor& self,
+                                          at::TensorList indices) {
+  std::vector<at::Tensor> result;
+  for (auto& index : indices) {
+    if (index.type().scalarType() == at::kByte) {
+      // The sizes of the ByteTensor mask must match the sizes of the
+      // corresponding dimensions in self.
+      for (int64_t j = 0; j < index.dim(); j++) {
+        int64_t src_idx = result.size() + j;
+        XLA_CHECK_EQ(index.size(j), self.size(src_idx))
+            << "The shape of the mask " << index.sizes() << " at index " << j
+            << " does not match the shape of the indexed tensor "
+            << self.sizes() << " at index " << src_idx;
+      }
+      // Replace with nonzeros.
+      auto nonzero = index.nonzero();
```

Comment:
If the majority of use cases pass in Byte tensors, we did all this for nothing (or almost so) :|

Reply:
That's a distinct possibility, could still be a win if we just do the `nonzero` on the host and finish the gather on device. But yeah, not great.